Give CValidationInterface Support for calling notifications on the CScheduler Thread

TheBlueMatt opened the pull request:
Apologies for the plumbing-only no-changes PR, but the "move wallet updates out of cs_main into a background thread" changeset is a bit big for all one go, so instead I'm doing this first. It simply gives CValidationInterface the neccessary plumbing to handle callbacks on the CScheduler thread. See https://github.com/TheBlueMatt/bitcoin/commit/4e82e409fbf3473f94dfa8cffe2674e853d59054 for a commit which switches a callback into the background thread.

The CScheduler thread was super lonely, so I decided to use that instead of adding new threads.

This conflicts trivially with #10178, but not enough to avoid doing parallel reviews. After that and this are merged, "move wallet callbacks into background thread" is just one more (somewhat large, sadly) PR away. See https://github.com/TheBlueMatt/bitcoin/commits/2017-01-wallet-cache-inmempool for the full branch.

TheBlueMatt committed:
Make ValidationInterface signals-type-agnostic

(by hiding boost::signals stuff in the .cpp)

This allows us to give it a bit more intelligence as we move
forward, including routing some signals through CScheduler. While
the introduction of a "internals" pointer in the class is pretty
ugly, the fact that we no longer need to include boost/signals
directly from validationinterface.h is very much worth the loss.

TheBlueMatt committed:
Give CMainSignals a reference to the global scheduler

...so that it can run some signals in the background later

TheBlueMatt committed:
Add default arg to CScheduler to schedule() a callback now

TheBlueMatt committed:
Support more than one CScheduler thread for serial clients

This will be used by CValidationInterface soon.

This requires a bit of work as we need to ensure that most of our
callbacks happen in-order (to avoid synchronization issues in
wallet) - we keep our own internal queue and push things onto it,
scheduling a queue-draining function immediately upon new
callbacks.

TheBlueMatt commented:
Rebased after ##10178 merge.

morcos commented on src/validationinterface.cpp:
```diff
@@ -4,49 +4,172 @@
+    // our own queue here :(
+    CCriticalSection cs_callbacksPending;
+    std::list<std::function<void (void)>> callbacksPending;
+    std::atomic<bool> fCallbacksRunning = ATOMIC_VAR_INIT(false);
```
nit: i'm the last person who should comment on coding style, but isn't this easier to read:
`std::atomic<bool> fCallbacksRunning(false)`

TheBlueMatt replied:
My C++-fu is not good enough to get that to compile. Not actually sure why, though, frankly.

sipa replied:
Could you try
```c++
std::atomic<bool> fCallbacksRunning = std::atomic<bool>(false);
```
?

TheBlueMatt replied:
Yes, that works, but I think thats worse than using the init wrapper, which is what that wrapper exists for.

theuni replied:
Use braced initialization in a class/struct definition:
```c++
std::atomic<bool> fCallbacksRunning{false};
```

TheBlueMatt replied:
Rewrote to just use a regular bool inside the lock, no reason to really have an atomic here.

morcos commented:
fast review utACK

laanwj commented:
utACK e5d0c66 - are you planning on adding tests for this functionality later?

TheBlueMatt commented:
I don't currently have tests written, but given nearly everything passes through I've found its pretty well-excersized by existing wallet functional tests. I'll add unit tests for this to my to-do list.

On April 25, 2017 8:55:34 AM EDT, "Wladimir J. van der Laan" <notifications@github.com> wrote:
>utACK e5d0c66 - are you planning on adding tests for this functionality
>later?

laanwj commented on src/validationinterface.cpp:
```diff
@@ -4,49 +4,172 @@
+                callback = callbacksPending.front();
+                callbacksPending.pop_front();
+            }
+            callback();
```
This doesn't seem exception-safe. If an exception is raised inside here, `fCallbacksRunning` will stay set forever.

TheBlueMatt replied:
Indeed. Put a generic try {} catch(...) {Log} around it, not sure what else to do there.

laanwj replied:
You could catch the exception, clear the flag, and `throw` again. Though RAII is usually the best way in C++ to cover all exit paths.

TheBlueMatt replied:
OK, made a local class that will RAII it :)

sipa commented on src/validationinterface.cpp:
```diff
@@ -4,49 +4,177 @@
+    }
+
+    void AddToProcessQueue(const std::function<void (void)>& func) {
+        if (!scheduler)
```
That's scary, it would lose a callback. Shouldn't it assert in this case?

TheBlueMatt replied:
assert()ed

sipa commented on src/validationinterface.cpp:
```diff
@@ -4,49 +4,177 @@
+            return;
+        }
+
+        while (true) {
```
Instead of looping until there are no more scheduled callbacks, isn't easier/better to return, but reschedule itself? That way a large set of queued callbacks won't prevent the scheduler from running other (non-callback) jobs in the mean time.

TheBlueMatt replied:
Done.

sipa reviewed (commented):
I'll give you the same comment as you gave on #10148... you're adding unused and untested functionality.

The first commit seems a useful refactoring that could be its own PR, but the rest seem to be preparations for something that doesn't exist yet.

TheBlueMatt commented on src/validationinterface.cpp:
```diff
@@ -4,49 +4,177 @@
+}
+
+CMainSignals::~CMainSignals() {
+    delete internals;
```
Cant, sadly, as the whole point was to not make CMainSignalsInterface sit in the .h to avoid having a bost/signals include in half our codebase.

sipa replied:
`std::unique_ptr` seems to work fine with forward-declared types: https://github.com/sipa/bitcoin/commit/a4ecaad

TheBlueMatt replied:
Well, I'll be! Took your patch and squashed (hope you dont mind).

sipa replied:
Sorry, but I'm going to insist that you maintain this commit intact (nevermind that it's likely the exact same patch you'd come up with if I hadn't shown you) forever in your PR history. Note: sarcasm.

TheBlueMatt committed:
Use TestingSetup to DRY qt rpcnestedtests

TheBlueMatt commented:
@sipa re: unused and untested code, see #10286, which I think it would be reasonable to require at least concept acks on prior to merging this.

ryanofsky commented on src/validationinterface.cpp:
```diff
@@ -22,6 +27,64 @@ struct CMainSignalsInstance {
+    // We are not allowed to assume the scheduler only runs in one thread,
+    // but must ensure all callbacks happen in-order, so we end up creating
+    // our own queue here :(
+    CCriticalSection cs_callbacksPending;
```
In commit "Handle more than one CScheduler thread in CValidationInterface"

This seems like it would a lot simpler it just spawned a single thread to run the callbacks in sequence instead of relying on CScheduler and then doing a bunch of synchronization on top of it to keep things scheduled and prevent more than one callback from executing at the same time.

Also, I think it would be nice if this functionality were implemented in a standalone class that could be unit tested.

Anyway, implementation seems fine, I couldn't find any bugs.

TheBlueMatt replied:
Moved to scheduler.h, testing left as an excersize for the reader :p.

ryanofsky reviewed (commented):
utACK 4fceec59372b778f9966485f2a13f97e21b1db4c

ryanofsky commented on src/validationinterface.cpp:
```diff
@@ -22,6 +27,64 @@ struct CMainSignalsInstance {
+        callback();
+    }
+
+    void AddToProcessQueue(const std::function<void (void)>& func) {
```
In commit "Handle more than one CScheduler thread in CValidationInterface":

Should pass `func` by value or rvalue reference so you can move it into the callbacksPending list without a copy.

TheBlueMatt replied:
Done.

ryanofsky commented on src/validationinterface.cpp:
```diff
@@ -22,6 +27,64 @@ struct CMainSignalsInstance {
+            if (callbacksPending.empty()) return;
+            fCallbacksRunning = true;
+
+            callback = callbacksPending.front();
```
In commit "Handle more than one CScheduler thread in CValidationInterface"

Could add `std::move` here.

TheBlueMatt replied:
Done.

TheBlueMatt commented on src/validationinterface.h:
```diff
@@ -32,52 +31,63 @@ void UnregisterAllValidationInterfaces();
+struct CMainSignalsInstance;
+class CMainSignals {
+private:
+    std::unique_ptr<CMainSignalsInstance> internals;
```
Yea, the real motivation there was to not have to #include boost signals garbage everywhere, nothing more.

ryanofsky commented:
utACK 643a988101f12581725d08d71cfcb5d932a062ac. I like the new ProcessQueue location and std::move additions.

TheBlueMatt commented:
Squashed and tweaked commit wording.

ryanofsky commented:
utACK 830c26b55a2b4bc93509d3fa7f17a0b383a95835. Only change aside from the history squashing is switching to the schedule() default arg.

sipa commented:
utACK. Since you're adding a new class, would you mind making a few changes to match the new naming rules in the style guide?

TheBlueMatt commented:
Removed C prefix from classes without rebasing.

ryanofsky reviewed (commented):
utACK b04b156d31787884d4c1fd4808b3f1f397d45c04. Only changes were removing "C", adding "_m"

TheBlueMatt commented:
Very minor rebase.

ryanofsky reviewed (commented):
utACK 608fad0e17d47d0943ff5a9425c65a5ea4eeba6f. No changes since last review except fixes to conflicting #include lines.

ryanofsky commented:
Status of this PR? It has had 4 utACKs (though @laanwj's is in [strikethrough](https://github.com/bitcoin/bitcoin/pull/10179#issuecomment-297020818)) as well as comments from @mchrostowski in #10286.

sipa commented on src/scheduler.h:
```diff
@@ -79,4 +81,26 @@ class CScheduler
+
+    CCriticalSection m_cs_callbacksPending;
+    std::list<std::function<void (void)>> m_callbacksPending;
+    bool m_fCallbacksRunning = false;
```
Would you mind changing this to lowercase-only variable names (e.g. `m_cs_callbacks_pending`, `m_callbacks_pending`, `m_callbacks_running`)?

TheBlueMatt replied:
Done.

ryanofsky reviewed (commented):
utACK cdf4cd8f54239bab6d5658ec6eec6c531f34091c. Just naming style changes since last review.

sipa commented:
@laanwj Does the strikethrough utACK means you still have a concern?

theuni commented on src/validationinterface.cpp:
```diff
@@ -21,22 +26,23 @@ struct MainSignalsInstance {
 
 void CMainSignals::UnregisterBackgroundSignalScheduler() {
-    m_internals->m_scheduler = NULL;
+    m_internals.reset(nullptr);
```
This potentially leaves events in the SingleThreadedSchedulerClient process queue with a dangling pointer to "this", no? I believe SingleThreadedSchedulerClient needs interrupt/stop functionality.

TheBlueMatt replied:
Interrupt() will stop the processing of events long before we ever call UnregisterBackgroundSignalScheduler in init/bitcoind.cpp/qt/bitcoin.cpp. There isnt really a great way to solve this, I think, without just making the CScheduler own the thread(/group), but I went ahead and pushed a commit which will just call any remaining callbacks when Unregister is called.

ryanofsky replied:
In commit "Support more than one CScheduler thread for serial clients"

It'd be good to just reset the scheduler pointer here instead of going overboard and destroying all the boost signals at this point as well. It just seems like a random and unexpected thing to be doing in a function called `UnregisterBackgroundSignalScheduler` especially given new flush behavior which allows signals be forwarded after the scheduler is stopped.

TheBlueMatt committed:
Flush CValidationInterface callbacks prior to destruction

Note that the CScheduler thread cant be running at this point,
it has already been stopped with the rest of the init threadgroup.
Thus, just calling any remaining loose callbacks during Shutdown()
is sane.

laanwj commented:
@sipa Just that I had a concern at the time, and was too quick to utACK. I should re-review.

laanwj commented on src/validationinterface.h:
```diff
@@ -32,52 +32,67 @@ void UnregisterAllValidationInterfaces();
+    virtual void BlockChecked(const CBlock&, const CValidationState&) {}
     /** Notifies listeners that a key for mining is required (coinbase) */
-    boost::signals2::signal<void (std::shared_ptr<CReserveScript>&)> ScriptForMining;
+    virtual void GetScriptForMining(std::shared_ptr<CReserveScript>&) {};
```
#10683 gets rid of `GetScriptForMining` in validationinterface, which I think it's good because it mutates its argument. Doing that in a decoupled thread would be disastrous.

TheBlueMatt replied:
All of the move-to-background-thread stuff is going to be a slow per-function process. See, eg, #10652 which moves a few functions to the background for net_processing and, of course, #10286 which moves a few to the background for wallet. Agreed that GetScriptForMining should go away and then it wont be a concern :).

theuni commented:
utACK (after rebase) now that #10683 is in.

TheBlueMatt commented:
Rebased.

TheBlueMatt commented:
Redid the shutdown callback-flushing at @morcos's request - now flushes earlier in the Shutdown() process in a much better location - after all our peers have stopped processing (and thus cant generate callbacks) and before wallet flushing.

TheBlueMatt commented:
Added a comment on flush/drop behavior - 
"Any future callbacks will be dropped. This should absolutely be safe - if missing a callback results in an unrecoverable situation, unclean shutdown would too. The only reason to do the above flushes is to let the wallet catch up with our current chain to avoid any strange pruning edge cases and make next startup faster by avoiding rescan."

morcos commented:
utACK 3192975

theuni commented:
re-utACK 3192975f1d177aa9f0bbd823c6387cfbfa943610, though very hesitantly after thinking through the tear-down some more. The shutdown process is going to get really hard to trace when callbacks start coming in on different threads.

I really hope some ownership model will begin to emerge.

ryanofsky commented on src/scheduler.h:
```diff
@@ -101,6 +101,9 @@ class SingleThreadedSchedulerClient {
     SingleThreadedSchedulerClient(CScheduler *pschedulerIn) : m_pscheduler(pschedulerIn) {}
     void AddToProcessQueue(std::function<void (void)> func);
+
+    // Processes all remaining queue members on the calling thread, blocking until queue is empty
```
In commit "Flush CValidationInterface callbacks prior to destruction"

Comment should point out that this should only be called when scheduler is not active. Otherwise this could burn cpu racing with scheduler thread to run the next callback.

TheBlueMatt replied:
Added the assert instead

ryanofsky reviewed (commented):
utACK 3192975f1d177aa9f0bbd823c6387cfbfa943610. Changes since last review were rebase after multiwallet & ScriptForMining removal, and adding the flush commit.

I left a few suggestions pertaining to the flush commit. I think this code would be good to clean up at some point, though it don't think it should hold up the PR.

TheBlueMatt commented on src/init.cpp:
```diff
@@ -251,6 +251,7 @@ void Shutdown()
     }
 #endif
     UnregisterAllValidationInterfaces();
+    GetMainSignals().UnregisterBackgroundSignalScheduler();
```
We cant unregister the background scheduler from the validation interface until we're sure nothing is gonna generate anymore callbacks (so, really, after the pcoinsTip/etc deletions). If we want to mirror the initialization order, we'd have to move it even further down, not up, as de-init in RAII order would be after wallet deltion.

TheBlueMatt committed:
Expose if CScheduler is being serviced, assert its not in EmptyQueue

TheBlueMatt commented on src/validationinterface.cpp:
```diff
@@ -44,6 +44,10 @@ void CMainSignals::UnregisterBackgroundSignalScheduler() {
 }
 
+void CMainSignals::FlushBackgroundCallbacks() {
+    m_internals->m_schedulerClient.EmptyQueue();
```
Added.
